pkg/kernel/networkservice/connectioncontextkernel/ipcontext/ipneighbors/common.go

```diff
@@ -45,31 +54,97 @@ func create(conn *networkservice.Connection) error {
 			return errors.WithStack(err)
 		}
 
-		if err = netlinkHandle.LinkSetUp(l); err != nil {
+		if err := setIPContextNeighbors(ctx, netlinkHandle, conn.GetContext().GetIpContext().GetIpNeighbors(), l); err != nil {
 			return errors.WithStack(err)
 		}
 
-		if err := setIPNeighbors(conn.GetContext().GetIpContext().GetIpNeighbors(), l); err != nil {
-			return errors.WithStack(err)
+		// If payload is IP - we need to add additional neighbor
+		if conn.GetPayload() == payload.IP {
+			peerLink, ok := peer.Load(ctx, isClient)
+			if !ok {
+				log.FromContext(ctx).Error("Peer link not found")
+				return nil
+			}
+			if peerLink == nil || peerLink.Attrs() == nil || peerLink.Attrs().HardwareAddr == nil {
+				panic(fmt.Sprintf("unable to construct peer ip neighbor %+v", peerLink))
+			}
+
+			dstNets := conn.GetContext().GetIpContext().GetDstIPNets()
+			if isClient {
+				dstNets = conn.GetContext().GetIpContext().GetSrcIPNets()
+			}
+
+			for _, dstNet := range dstNets {
+				if dstNet != nil {
+					if err := setPeerNeighbor(ctx, netlinkHandle, l, peerLink, dstNet); err != nil {
+						return err
+					}
+				}
+			}
 		}
 	}
 	return nil
 }
 
-func setIPNeighbors(ipNeighbours []*networkservice.IpNeighbor, netLink netlink.Link) error {
+func setIPContextNeighbors(ctx context.Context, handle *netlink.Handle, ipNeighbours []*networkservice.IpNeighbor, netLink netlink.Link) error {
 	for _, ipNeighbor := range ipNeighbours {
 		macAddr, err := net.ParseMAC(ipNeighbor.HardwareAddress)
 		if err != nil {
 			return errors.Wrapf(err, "invalid neighbor MAC address: %v", ipNeighbor.HardwareAddress)
 		}
-		if err := netlink.NeighAdd(&netlink.Neigh{
+		now := time.Now()
+		neigh := &netlink.Neigh{
 			LinkIndex:    netLink.Attrs().Index,
 			State:        link.NudReachable,
 			IP:           net.ParseIP(ipNeighbor.Ip),
 			HardwareAddr: macAddr,
-		}); err != nil && !os.IsExist(err) {
+		}
+		if err := handle.NeighSet(neigh); err != nil {
```

Comment:
Was the root of the bug the use of NeighAdd vs NeighSet?

Reply:
No, it is minor change, for uniformity.